fix(GH-3574): Added root execution context message headers support

igdianov opened the pull request:
This PR adds support for injecting root execution context attributes into engine events message headers. The producer and consumers will be able inspect message headers to correlate, route, and filter message events with the following root exection context headers if available:

- [x] rootProcessInstanceId
- [x] rootProcessDefinitionId
- [x] rootProcessDefinitionKey
- [x] rootProcessDefinitionVersion
- [x] deploymentId
- [x] deploymentName
- [x] deploymentVersion

Optional root execution context may include:

- [x] rootBusinessKey
- [x] rootProcessName
- [x] rootProcessDefinitionName

This can be used to enable support for Spring Cloud Stream partitioned query and audit consumers with consistent message routing between partitions using rootProcessInstanceId hash value , i.e.

```
# split engine events stream to 10 partitions
spring.cloud.stream.bindings.auditProducer.producer.partition-count=10

# Use rootProcessInstanceId header to route message between partitions or use random hash value if missing header
spring.cloud.stream.bindings.auditProducer.producer.partition-key-expression=headers['rootProcessInstanceId']?:T(java.lang.Math).random()*100
```

Fixes https://github.com/Activiti/Activiti/issues/3574

igdianov committed:
fix(AAE-731): Added root execution context message headers support

erdemedeiros committed:
Fix typo

erdemedeiros reviewed (changes requested):
@igdianov, as discussed earlier, it's probably better to make it explicitly in the header name that's the "root" execution. Also having an IT testing the headers with a Call Activity would be nice.

igdianov committed:
fix: add root prefix to execution context message headers

igdianov committed:
fix: rename APP_VERSION header name to DEPLOYMENT_VERSION

igdianov committed:
fix: add IT coverage for ExecutionContextMessageHeader with subprocess

igdianov commented:
> @igdianov, as discussed earlier, it's probably better to make it explicitly in the header name that's the "root" execution. Also having an IT testing the headers with a Call Activity would be nice.

@erdemedeiros I have fixed the message header names and added test coverage for execution context message header values for execution with subprocesses.

erdemedeiros reviewed (approved):
LGTM!
